src/gmcp.cpp

```diff
@@ -0,0 +1,404 @@
+  json j;
+
+  j["physical"] = GET_PHYSICAL(ch);
+  j["mental"] = GET_MENTAL(ch);
+  j["physical_max"] = GET_MAX_PHYSICAL(ch);
+  j["mental_max"] = GET_MAX_MENTAL(ch);
+  j["karma"] = GET_KARMA(ch);
+  j["nuyen"] = GET_NUYEN(ch);
+  j["carrying_weight"] = IS_CARRYING_W(ch);
+  j["max_carry_weight"] = CAN_CARRY_W(ch);
+  j["armor"] = json::object();
+  j["armor"]["ballistic"] = GET_BALLISTIC(ch);
+  j["armor"]["impact"] = GET_IMPACT(ch);
+  if (get_ch_domain_str(ch, TRUE)) 
+    j["domain"] = get_ch_domain_str(ch, TRUE);
+
+  // Current Ammo Handler
+  if (GET_EQ(ch, WEAR_WIELD) &&
+      IS_GUN(GET_OBJ_VAL(GET_EQ(ch, WEAR_WIELD), 3)))
+    if (GET_EQ(ch, WEAR_WIELD)->contains) {
+      j["ammo"] = MIN(GET_OBJ_VAL(GET_EQ(ch, WEAR_WIELD), 5), GET_OBJ_VAL(GET_EQ(ch, WEAR_WIELD)->contains, 9));
+    } else
+      j["ammo"] = 0;
+    else
+      j["ammo"] = 0;
+
+  // Max ammo handler
+  if (GET_EQ(ch, WEAR_WIELD) &&
+      IS_GUN(GET_OBJ_VAL(GET_EQ(ch, WEAR_WIELD), 3)))
+    j["max_ammo"] =  GET_OBJ_VAL(GET_EQ(ch, WEAR_WIELD), 5);
+  else
+    j["max_ammo"] = 0;
+
+  // Dump the json to a string and send it.
+  std::string payload = j.dump();
+  SendGMCP(ch->desc, "Char.Vitals", payload.c_str());
+}
+
+void SendGMCPCharPools( struct char_data * ch )
+{
+  if (!ch || !ch->desc || !ch->desc->pProtocol->bGMCP) return;
+  json j;
+
+  j["pools"] = json::object();
+  j["pools"]["combat"] = json::object();
+  j["pools"]["combat"]["body"] = GET_BODY_POOL(ch);
+  j["pools"]["combat"]["dodge"] = GET_DODGE(ch);
+  j["pools"]["combat"]["offense"] = GET_OFFENSE(ch);
+  j["pools"]["combat"]["max"] = GET_COMBAT_POOL(ch);
+
+  j["task_pools"] = json::object();
+  for (int x = 0; x < 7; x++) {
+    j["task_pools"][attributes[x]] = GET_TASK_POOL(ch, x);  
+  }
+
+  if (GET_ASTRAL(ch) > 0)
+    j["pools"]["astral"] = GET_ASTRAL(ch);
+  if (GET_HACKING(ch) > 0)
+    j["pools"]["hacking"] = GET_HACKING(ch);
+  if (GET_MAGIC_POOL(ch) > 0) {
+    j["pools"]["magic"] = json::object();
+    j["pools"]["magic"]["spell"] = GET_MAGIC_POOL(ch);
+    j["pools"]["magic"]["casting"] = GET_CASTING(ch);
+    j["pools"]["magic"]["drain"] = GET_DRAIN(ch);
+    j["pools"]["magic"]["defense"] = GET_SDEFENSE(ch);
+    if (GET_METAMAGIC(ch, META_REFLECTING) == 2)
+      j["pools"]["magic"]["reflecting"] = GET_REFLECT(ch);
+  }
+
+  if (GET_CONTROL(ch) > 0)
+    j["pools"]["control"] = GET_CONTROL(ch);
+
+  // Dump the json to a string and send it.
+  std::string payload = j.dump();
+  SendGMCP(ch->desc, "Char.Pools", payload.c_str());
+}
+
+
+void ExecuteGMCPMessage(descriptor_t *apDescriptor, const char *module, const json &payload) {
+  if (!strncmp(module, "Core.Supports.Get", strlen(module)))
+    SendGMCPCoreSupports(apDescriptor);
+  else if (!strncmp(module, "Room.Info.Get", strlen(module)))
+    if (!apDescriptor->character)
+      mudlog_vfprintf(NULL, LOG_SYSLOG, "Unable to Handle GMCP Room.Info.Get Call: No Valid Character");
```

Comment:
All of these calls should be `apDescriptor->character` instead of NULL. The code knows how to handle null character pointers.